File: android/app/src/main/java/com/dungeonboss/game/CrawlResolver.kt
package com.dungeonboss.game

import com.dungeonboss.model.Encounter
import com.dungeonboss.model.Hero

/**
 * Runs a single hero through a single dungeon and reports the outcome. Applies
 * room/boss damage reduced by the hero's special ability. Stateless; returns a
 * result rather than mutating anything.
 */
object CrawlResolver {
    enum class Outcome { DIED, ESCAPED }

    /** One step: the [damage] is the amount actually taken (after the ability). */
    data class Step(val encounter: Encounter, val damage: Int, val healthAfter: Int)

    data class Result(val outcome: Outcome, val remainingHealth: Int, val log: List<Step>)

    fun resolve(hero: Hero, dungeon: Dungeon): Result {
        val ability = HeroAbility.lookup(hero)
        var health = hero.health
        val log = mutableListOf<Step>()

        for (encounter in dungeon.encounters()) {
            val damage = ability.reduced(encounter, encounter.damage)
            health -= damage
            log.add(Step(encounter, damage, health))
            if (health <= 0) break
        }

        val outcome = if (health <= 0) Outcome.DIED else Outcome.ESCAPED
        return Result(outcome, health, log)
    }
}

File: webapp/lib/crawl_resolver.rb
# frozen_string_literal: true

require_relative "hero_ability"

# Runs a single hero through a single dungeon and reports the outcome.
# Applies room/boss damage reduced by the hero's special ability (HeroAbility).
# Stateless; returns a result rather than mutating anything.
class CrawlResolver
  # The result of one crawl.
  # outcome: :died or :escaped
  # remaining_health: hero health after the crawl (<= 0 if died)
  # log: ordered list of { encounter:, damage:, health_after: } steps, where
  #      damage is the amount actually taken (after the hero's ability).
  Result = Struct.new(:outcome, :remaining_health, :log, keyword_init: true)

  # @param hero [Hero]
  # @param dungeon [Dungeon]
  # @return [Result]
  def self.resolve(hero, dungeon)
    ability = HeroAbility.lookup(hero)
    health = hero.health
    log = []

    dungeon.encounters.each do |encounter|
      damage = ability.reduced(encounter, encounter.damage)
      health -= damage
      log << { encounter: encounter, damage: damage, health_after: health }
      break if health <= 0
    end

    outcome = health <= 0 ? :died : :escaped
    Result.new(outcome: outcome, remaining_health: health, log: log)
  end
end
